### PR TITLE
Finding loci via a distance based clumping

### PR DESCRIPTION
**Pull request by DSuveges**
When given a set of single point associations, lead snps are identified using windows of a given size, where the distance between leads cannot be smaller than the window size. 

- The algorithm orders positions by p-value, 
- Then finds the most significant association
- Excludes all further associations from testing where the position is less than the window size from the found lead.
- Repeats until no variants left.

The function returns with a study-locus object

### CODE REVIEW

**Comment by d0choa** [comment]
I added some updates:
- Minor bugfix from [465306d](https://github.com/opentargets/genetics_etl_python/pull/97/commits/465306d79490057d41a3a102dec7c92f4c2b93d5)
- Testing increased to 100% coverage of the new method (semantic testing of the `clump` function is not implemented)
- `Window-based clumping` added to the documentation
- `SummaryStatistics` now has a function to clump using the new method. Tests for the new function were also added.

The result is that the tests are failing because the output of the `clump` function doesn't yet look like a `StudyLocus`. If I'm not missing anything this is the only thing pending.

File: src/otg/method/window_based_clumping.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+                ),
+            )
+            .groupBy("cluster_id")
+            # Aggregating all data from each cluster:
+            .agg(
+                WindowBasedClumping._collect_clump(
+                    f.col("pValueMantissa"), f.col("pValueExponent")
+                ).alias("clump")
+            )
+            # Explode and extract columns:
```
Comment (ireneisdoomed) [inline src/otg/method/window_based_clumping.py:325]:
```suggestion
            # Explode and identify the index variant representative of the cluster:
```

**Review by ireneisdoomed — COMMENTED** [review]
Amazing. Really intricate algorithm. I've made some minor suggestions.
I have some general ones, lmk your thoughts:
- I'd like to have a `method/clumping.py` module with the distance based and the LD based (which currently lives inside `ld.py`).
- This method is very summary stats focused, for obvious reasons. However, is it true that it could be applied to a "simple" StudyLocus dataset? For that to happen, the `clump` class method should accept a SummaryStatistics or a StudyLocus object.
- `clump` does not return a StudyLocus dataset yet. Currently the structure returns something like:
````
root
 |-- studyLocusId: string (nullable = true)
 |-- cluster_id: string (nullable = true)
 |-- position: long (nullable = true)
 |-- pValueMantissa: double (nullable = true)
 |-- pValueExponent: long (nullable = true)
````
I assume that this needs should be inner joined with the original SummaryStats/StudyLocus object.

File: src/otg/method/window_based_clumping.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+                        f.lit(window_length),
+                    )
+                ),
+                lambda x, y: f.when(y == 1.0, x.withField("isLead", y)),
+            ),
+            lambda col: col.isNotNull(),
+        )
+
+    @staticmethod
+    def _collect_clump(mantissa: Column, exponent: Column) -> Column:
```
[the remark this answers is not in the document]
Comment (d0choa) [reply src/otg/method/window_based_clumping.py:239]: sounds good

File: src/otg/method/window_based_clumping.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+            summary_stats (SummaryStatistics): summary statistics to clump
+            window_length (int): window length in basepair
+
+        Returns:
+            StudyLocus: clumped summary statistics
+        """
+        return StudyLocus(
+            _df=summary_stats.df.withColumn(
+                "cluster_id",
+                WindowBasedClumping._identify_cluster_peaks(
```
[the remark this answers is not in the document]
Comment (d0choa) [reply src/otg/method/window_based_clumping.py:311]: add the comment no need to discuss it I believe

**Review by d0choa — APPROVED** [review]

**Pull request merged**